File: backend/src/routes/specialDays.js
import express from "express";
import { requireAuth, requireRole } from "../middleware/auth.js";
import { ROLES } from "../constants/roles.js";
import { SPECIAL_DAY_TYPES } from "../constants/specialDayTypes.js";
import {
    getAllSpecialDays,
    createSpecialDay,
    updateSpecialDay,
    deleteSpecialDay,
} from "../services/specialDayService.js";
import { broadcast } from "../services/sseService.js";

const router = express.Router();

const ALLOWED_ROLES = [ROLES.ADMIN, ROLES.LOGISTICS];
const validTypes = Object.values(SPECIAL_DAY_TYPES);

/**
 * GET /api/special-days?month=YYYY-MM
 * Optional month filter. If omitted, returns all.
 */
router.get("/", requireAuth, requireRole(ALLOWED_ROLES), (req, res) => {
    let days = getAllSpecialDays();
    const month = req.query.month;
    if (month) {
        days = days.filter((d) => d.date.startsWith(month));
    }
    res.json(days);
});

router.post("/", requireAuth, requireRole(ALLOWED_ROLES), (req, res) => {
    const { date, type, note, meals } = req.body;

    if (!date || !type) {
        return res.status(400).json({ error: "date and type are required" });
    }

    if (!validTypes.includes(type)) {
        return res.status(400).json({ error: `Invalid type. Must be one of: ${validTypes.join(", ")}` });
    }

    try {
        const entry = createSpecialDay({
            date,
            type,
            note,
            meals,
            createdBy: req.session.user.id,
        });
        broadcast("special-day-change", { date });
        res.status(201).json(entry);
    } catch (err) {
        res.status(409).json({ error: err.message });
    }
});

router.put("/:date", requireAuth, requireRole(ALLOWED_ROLES), (req, res) => {
    const { date } = req.params;
    const { type, note, meals } = req.body;

    if (type && !validTypes.includes(type)) {
        return res.status(400).json({ error: `Invalid type. Must be one of: ${validTypes.join(", ")}` });
    }

    try {
        const updated = updateSpecialDay(date, { type, note, meals });
        broadcast("special-day-change", { date });
        res.json(updated);
    } catch (err) {
        res.status(404).json({ error: err.message });
    }
});

router.delete("/:date", requireAuth, requireRole(ALLOWED_ROLES), (req, res) => {
    try {
        deleteSpecialDay(req.params.date);
        broadcast("special-day-change", { date: req.params.date });
        res.json({ message: "Deleted" });
    } catch (err) {
        res.status(404).json({ error: err.message });
    }
});

export default router;

File: mhp-python/app/routes/special_days.py
from fastapi import APIRouter, Depends, HTTPException

from app.auth import require_roles
from app.constants import Roles, SpecialDayTypes
from app.models import SpecialDayRequest
from app.services.special_day_service import (
    get_special_days_for_month, create_special_day, update_special_day, delete_special_day, get_special_day,
)

router = APIRouter()

_ALLOWED = [Roles.ADMIN, Roles.LOGISTICS]


@router.get("")
def list_special_days(month: str = None, user: dict = Depends(require_roles(_ALLOWED))):
    from datetime import date as DateType
    m = month or DateType.today().strftime("%Y-%m")
    return get_special_days_for_month(m)


@router.post("")
def create(body: SpecialDayRequest, user: dict = Depends(require_roles(_ALLOWED))):
    if body.type not in (SpecialDayTypes.OFFICE_CLOSED, SpecialDayTypes.GOVT_HOLIDAY, SpecialDayTypes.CELEBRATION):
        raise HTTPException(status_code=400, detail="Invalid special day type")

    existing = get_special_day(body.date)
    if existing:
        raise HTTPException(status_code=400, detail=f"A special day already exists for {body.date}")

    data = body.model_dump(exclude_none=True)
    data["createdBy"] = user["id"]
    return create_special_day(data)


@router.put("/{date}")
def update(date: str, body: SpecialDayRequest, user: dict = Depends(require_roles(_ALLOWED))):
    result = update_special_day(date, body.model_dump(exclude_none=True))
    if not result:
        raise HTTPException(status_code=404, detail=f"No special day found for {date}")
    return result


@router.delete("/{date}")
def delete(date: str, user: dict = Depends(require_roles(_ALLOWED))):
    if not delete_special_day(date):
        raise HTTPException(status_code=404, detail=f"No special day found for {date}")
    return {"message": f"Special day {date} deleted"}
